### PR TITLE
Add Header Only API `pairwise_linestring_intersection`

### LINKED ISSUE #765: header-only implementation of intersection
Implement header only API for `intersection_count` and `intersection`. 

```c++
template<typename Range1, typename Range2, typename OutputIt>
OutputIt pairwise_linestring_intersection(Range1 multilinestrings1, Range2 multilinestring2, OutputIt output_begin, stream);

template<typename Range1, typename Range2, typename OutputIt>
OutputIt pairwise_linestring_intersection_count(Range1 multilinestring1, Range2 multilinestring2, OutputIt output_begin, stream);
```

Depending on #763 and #764. With the help of these primitive and data structures, I posturizes that the kernel should be parallelized on 1-segment per thread and iterate on the multilinestring in the other multilinestring.

Kernel pseudo code without needing to write indices 

```c++
__global__ void intersection_count_kernel(range1, range2, output):
	tidx = get_threadidx()
	Segment seg1 = range1.segment(tidx)
	For (mls : range2)
		For (seg2 : mls)
			Point_opt, line_opt = segment_intersection(seg1, seg2)
			If (point_opt != nullopt)
				atomicInc(Output[tidx].num_intersecting_points)
			If (line_opt != nullopt)
				atomicInc(Output[tidx].num_overlapping_segments)
```

```c++
__global__ void intersection_kernel(range1, range2, temp_storage, output):
	tidx = get_threadidx()
	Segment seg1 = range1.segment(tidx)
	For (mls : range2)
		For (seg2 : mls)
			point_opt, line_opt = segment_intersection(seg1, seg2)
			If (point_opt != nullopt)
				Npoint = atomicReadAndInc(temp_storage.npoint)
				Output[npoint] = point_opt.value()
			If (line_opt != nullopt)
				NLine  = atomicReadAndInc(temp_storage.nline)
				Output[nLine] = line_opt.value()
```

[UPDATE: 11-8]
`count` API is only counting the upper bound of the number of intersection/overlaps - it does not give any promises to the number of results. So `count` will not be exposed as a public API.

https://github.com/rapidsai/cuspatial/issues/765#tasklist-block-52d7347e-d26a-4c76-968c-5e78b76839c8

### PR DESCRIPTION
**Pull request by isVoid**
## Description

This PR adds header only API `pairwise_linestring_intersection`. Closes #765 .

Depends on #813 #818 #819 #851

## Checklist
- [x] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cuspatial/blob/HEAD/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.

### CODE REVIEW

**Comment by thomcom** [comment]
Waiting for the dependencies to be merged before reviewing this.

File: cpp/include/cuspatial/experimental/detail/find/find_point_on_segments.cuh
Comment (harrism) [inline cpp/include/cuspatial/experimental/detail/find/find_point_on_segments.cuh:75]: Should be `find_points_on_segments`? multipoints (plural), and a set of segments (plural)

File: cpp/include/cuspatial/experimental/linestring_intersection.cuh
Comment (harrism) [inline cpp/include/cuspatial/experimental/linestring_intersection.cuh:81]: Does this result have duplicates? I thought this one had no duplicates.

**Review by harrism — APPROVED** [review]
This all came together really well. I can't see anything major wrong, just a couple of comments.

File: cpp/tests/experimental/spatial/linestring_intersection_test.cu
Comment (thomcom) [inline cpp/tests/experimental/spatial/linestring_intersection_test.cu:117]: Is this for this PR? Seems hard to compare `got` and `expected` if results are arbitrary?

File: cpp/tests/experimental/spatial/linestring_intersection_test.cu
Comment (thomcom) [inline cpp/tests/experimental/spatial/linestring_intersection_test.cu:171]: :thumb: No duplicates!

File: cpp/tests/experimental/spatial/linestring_intersection_test.cu
Comment (thomcom) [inline cpp/tests/experimental/spatial/linestring_intersection_test.cu:204]: The `ids` don't seem correct to me here, as they're all 0. `lhs` looks right, there's only one linestring and only one segment. On `rhs` however I'm confused at the output. There's only one linestring so 0 is correct for line #203, but line 204 is confusing. All three segments overlap, which you've combined into the result segment, but only segment 0 is returned.

If the user wanted for example to display each intersecting segment from `rhs` in different colors, this output would not enable that, I think? I'm ok with leaving the API as it is but this might need consideration.

File: cpp/tests/experimental/spatial/linestring_intersection_test.cu
Comment (thomcom) [inline cpp/tests/experimental/spatial/linestring_intersection_test.cu:231]: Does this not contain anything because the merged point exists in the input data, so the user can look it up?

File: cpp/tests/experimental/spatial/linestring_intersection_test.cu
Comment (thomcom) [inline cpp/tests/experimental/spatial/linestring_intersection_test.cu:230]: A similar problem exists here where the user cannot recover which points in `rhs` intersect.

File: cpp/tests/experimental/spatial/linestring_intersection_test.cu
Comment (thomcom) [reply cpp/tests/experimental/spatial/linestring_intersection_test.cu:231]: Actually shouldn't there be a return point, `P{0.5, 0.5}`  caused by the intersection when `({0.0, 1.0}, {1.0, 0.0})` crosses `({0.0, 0.0}, {3.0, 3.0})`? I see that got merged into `P{0.0, 2.0}`.

File: cpp/tests/experimental/spatial/linestring_intersection_intermediates_remove_if_test.cu
Comment (thomcom) [inline cpp/tests/experimental/spatial/linestring_intersection_intermediates_remove_if_test.cu:92]:
```suggestion
// float and double are logically the same but would require separate tests due to precision.
```

File: cpp/tests/experimental/find/find_point_on_segments_test.cu
Comment (thomcom) [inline cpp/tests/experimental/find/find_point_on_segments_test.cu:53]: All the `Simple` `FindPointOnSegmentTest`s test for flags `[0, 1]`. I'd like to see the basic set enumerated with these tests: `[0, 0]`, `[0, 1]`, `[1,0]`, `[1, 1]`. Ideally the same breakdown for `multipoints`

File: cpp/include/cuspatial_test/vector_factories.cuh
Comment (thomcom) [inline cpp/include/cuspatial_test/vector_factories.cuh:39]: I thought that `device_uvector` meant uninitialized. If so, `make_device_uvector` from an initialization list doesn't make sense to me as a name.

File: cpp/include/cuspatial/experimental/detail/linestring_intersection_with_duplicates.cuh
Comment (thomcom) [inline cpp/include/cuspatial/experimental/detail/linestring_intersection_with_duplicates.cuh:200]: I'd love to see more detailed comments about the steps in this function.

**Review by thomcom — CHANGES_REQUESTED** [review]
Looks really nice. Just a few comments about testing strategy and one about internal documentation. Most of my comments are just trying to understand the output of the final API. Do we have a concern about dropping duplicates and making it harder for the user to identify which points intersect or their count?

File: cpp/tests/experimental/spatial/linestring_intersection_intermediates_remove_if_test.cu
Comment (isVoid) [reply cpp/tests/experimental/spatial/linestring_intersection_intermediates_remove_if_test.cu:78]: The name of the method is modeled after thrust: https://nvidia.github.io/thrust/api/groups/group__stream__compaction.html#function-remove-if with similar input.

File: cpp/tests/experimental/spatial/linestring_intersection_test.cu
Comment (isVoid) [reply cpp/tests/experimental/spatial/linestring_intersection_test.cu:117]: Yeah, this still needs to be done. Otherwise the test maybe flaky.

File: cpp/include/cuspatial/experimental/detail/linestring_intersection_with_duplicates.cuh
Comment (isVoid) [reply cpp/include/cuspatial/experimental/detail/linestring_intersection_with_duplicates.cuh:200]: Done in #851

File: cpp/include/cuspatial/experimental/linestring_intersection.cuh
Comment (isVoid) [reply cpp/include/cuspatial/experimental/linestring_intersection.cuh:81]: No, it's a typo.

File: cpp/tests/experimental/spatial/linestring_intersection_test.cu
Comment (isVoid) [reply cpp/tests/experimental/spatial/linestring_intersection_test.cu:204]: I think the original requirement for intersection is a ambiguous and this is certainly not considered in original design. Can file a separate issue to track if desired.

File: cpp/tests/experimental/spatial/linestring_intersection_test.cu
Comment (isVoid) [reply cpp/tests/experimental/spatial/linestring_intersection_test.cu:231]: Yes, that's the purpose of https://github.com/rapidsai/cuspatial/pull/819

**Comment by isVoid** [comment]
> Do we have a concern about dropping duplicates and making it harder for the user to identify which points intersect or their count?

Unfortunately, the original requirement for the output of intersection is unclear about what the user really expects here. I went afar to return the most cleaned-up output that the algorithm can return. (So does geopandas) It could be an early-optimization. It could be likely that it's not what the user wants. If you think the design is not meeting the need, feel free to file an issue to track. 

My question is, does this intersection function help your binary predicate use case?

File: cpp/tests/experimental/spatial/linestring_intersection_test.cu
Comment (isVoid) [reply cpp/tests/experimental/spatial/linestring_intersection_test.cu:231]: Hmm, Shapely seems to merge the points, but not the segments:
```python
In [1]: from shapely.geometry import *

In [2]: l = LineString([(0.0, 0.0), (3.0, 3.0)])

In [3]: l2 = LineString([(0.0, 1.0), (1.0, 0.0), (2.0, 2.0), (0.0, 0.0)])

In [4]: l.intersection(l2)
Out[4]: <MULTILINESTRING ((0 0, 0.5 0.5), (0.5 0.5, 2 2))>
```

File: cpp/tests/experimental/spatial/linestring_intersection_test.cu
Comment (isVoid) [reply cpp/tests/experimental/spatial/linestring_intersection_test.cu:204]: To support your use case, you would need to group all merged geometry ids, this requires an additional level of offsets into the id arrays. I think that's a pretty big change and would take up a separate effort.

From a different angle, can't we pose this as a separate step from intersection? You have a intersection, finding where the intersection comes from is a search problem.

Comment (harrism) [reply cpp/tests/experimental/spatial/linestring_intersection_test.cu:204]: We may want to factor out the intersection metadata (segment indices) as a search problem in the future anyway.

File: cpp/include/cuspatial_test/vector_factories.cuh
Comment (harrism) [reply cpp/include/cuspatial_test/vector_factories.cuh]: I think it's fine. a device_uvector is a container that is never *default* initialized. But you can still put data in it. This factory function just creates a uvector and then copies data into it.  

Note this is only available in tests, it is not intended for use in the library or by users of the library.  I kind of agree that we could probably get away with just `make_device_vector` in tests...

Comment (isVoid) [reply cpp/include/cuspatial_test/vector_factories.cuh]: I think the main reason why I added this helper is because `intersection_result` consists of `device_uvector`. And I tried to keep things simple by constructing `device_uvector` in test and just copy/move it to the struct. Using `device_vector` will require some extra steps.

File: cpp/tests/experimental/spatial/linestring_intersection_test.cu
Comment (thomcom) [reply cpp/tests/experimental/spatial/linestring_intersection_test.cu:204]: I'm in favor of creating a new issue to track this, it doesn't need to block this PR.

**Comment by thomcom** [comment]
I'm sure this helps with implementing a great deal of binary predicates! I'm not exactly sure how, yet. I'm a little worried about a case of $\forall p in \lambda_1 \cap \lambda_2$ if it is required. This definitely covers `None` and `any` cases, which, without looking closely at dozens of intersection predicate calls yet, probably covers most of them.

**Comment by thomcom** [comment]
My questions are mostly general and for the purpose of understanding what's coming my way soon. :) I see @harrism already approved, I will approve if you want to get this merged before the week ends.

File: cpp/include/cuspatial_test/vector_factories.cuh
Comment (thomcom) [reply cpp/include/cuspatial_test/vector_factories.cuh]: I don't require a change, but was mostly referring to the fact that the function name doesn't compute well for me. Since `uvector` takes an initializer list and that users are going to understand that it takes an initializer list, I'm ok with the name.

File: cpp/tests/experimental/spatial/linestring_intersection_test.cu
Comment (thomcom) [reply cpp/tests/experimental/spatial/linestring_intersection_test.cu:231]: This result is really interesting to me. It seems that shapely combines the point intersection with the segment intersection in this example. It also seems like that's something that could be accomplished in python, though perhaps not as efficiently.

File: cpp/tests/experimental/spatial/linestring_intersection_test.cu
Comment (isVoid) [reply cpp/tests/experimental/spatial/linestring_intersection_test.cu:240]: Ok, I think sorting the result is finally added!

File: cpp/tests/experimental/spatial/linestring_intersection_test.cu
Comment (thomcom) [inline cpp/tests/experimental/spatial/linestring_intersection_test.cu:70]: I can see us not wanting this to be only a test utility down the road.

File: cpp/include/cuspatial_test/vector_factories.cuh
Comment (isVoid) [reply cpp/include/cuspatial_test/vector_factories.cuh]: I believe in c++ `make_*` is a convention for factory functions of classes. In standard c++, they were really meant to be a hack to make classes before CTAD. In cuDF, there are examples of building columns from various input specs that's non-standard to `cudf::column` ctor. I'm kinda abusing that idea here so it makes sense that it feels off. I'll add a line of comment to remind for better naming here.

**Review by thomcom — CHANGES_REQUESTED** [review]

File: cpp/tests/experimental/find/find_point_on_segments_test.cu
Comment (isVoid) [reply cpp/tests/experimental/find/find_point_on_segments_test.cu]: Done

File: cpp/include/cuspatial/experimental/detail/find/find_point_on_segments.cuh
Comment (isVoid) [reply cpp/include/cuspatial/experimental/detail/find/find_point_on_segments.cuh]: renamed

File: cpp/tests/experimental/spatial/linestring_intersection_test.cu
Comment (isVoid) [reply cpp/tests/experimental/spatial/linestring_intersection_test.cu:70]: #858

File: cpp/tests/experimental/spatial/linestring_intersection_test.cu
Comment (isVoid) [reply cpp/tests/experimental/spatial/linestring_intersection_test.cu:353]: #860

**Review by thomcom — APPROVED** [review]

**Comment by isVoid** [comment]
@gpucibot merge

**Pull request merged**
